### PR TITLE
Add Webhook debugging guide, updates to sync guide

### PR DESCRIPTION
**Pull request by royanger**
Please provide a high content/organization review. Once done and once changes are made I will proof this and re-request reviews.

Updated: https://docs-preview-851.clerkpreview.com/docs/integrations/webhooks/sync-data
Added: https://docs-preview-851.clerkpreview.com/docs/integrations/webhooks/debugging-webhooks

### CODE REVIEW

File: docs/integrations/webhooks/sync-data.mdx [hunk 1]
```diff
@@ -87,6 +87,24 @@ pnpm add svix
 
 </CodeBlockTabs>
 
+### Add your endpoint to Middleware
+
+Your Route Handler must be made public or ignored by Middleware to allow the request to succeed. The following example will make any webhooks created under `app/api/webhooks/` public. See [authMiddleware](/docs/references/nextjs/auth-middleware) for more information.
```
Comment (S3Prototype) [inline docs/integrations/webhooks/sync-data.mdx:92]:
```suggestion
Your API route must be made public or ignored by Middleware to allow the request to succeed. The following example will make any webhooks created under `app/api/webhooks/` or in `pages/api/webhooks.ts` public. See [authMiddleware](/docs/references/nextjs/auth-middleware) for more information.
```

Changed language to be inclusive of Pages router since we have a Pages example on the page

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 2]
```diff
@@ -0,0 +1,93 @@
+---
+title: Debugging webhooks
+description: Understand how to debug your webhooks while developing your application
+---
+
```
Comment (S3Prototype) [inline docs/integrations/webhooks/debugging-webhooks.mdx:5]:
```suggestion

# Debugging webhooks

```

H1 is required now. Frontmatter title is ignored

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+### Debugging your tunnel
+
+You've setup your route and tested it, configured your tunnel, updated your dashboard and now are trying to use the webhook but are getting errors. You will want to visit the [Webhook section](https://dashboard.clerk.com/last-active?path=webhooks) of your dashboard and the select the webhook you created.
+
+You will see a 'Webhook Attempts' section and likely see that one or more of those attempts have failed. Click on the failed attempt to expand it. In the details for the attempt, there will be a 'HTTP RESPONSE CODE'. The code will likely be either a 500 or something starting with a 4, like 401 or 404. See [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) from MDN for reference.
+
+IMAGES FROM THE WEBHOOK SECTION HERE
+
+Below is a table with some of the common response codes you might see, and what they indicate:
```
Comment (S3Prototype) [inline docs/integrations/webhooks/debugging-webhooks.mdx:65]: I think it's worth giving this its own H3 heading. "Webhook error response codes" for example. I can see someone wanting to link to this section, or jump straight to it from the TOC

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+When you are developing on your localhost, your application is not internet facing. It can't receive the webhook request. You will need to use a tool that creates a tunnel from the internet to your localhost. These tools will provide temporary or permanant URLs depending on the tool and the plan you subscribe to. Popular tools included `ngrok`, `localtunnel` and `Cloudflare Tunnel`.
+
+Below is an image demonstrating the flow from when you perform an action that triggers a webhook through until the webhook is received by your application
+
+FLOW CHART IMAGE HERE
+
+
+## Debugging
+
+With the Webhook flow in mind, the following areas could cause problems when trying to develop with webhooks:
```
Comment (S3Prototype) [inline docs/integrations/webhooks/debugging-webhooks.mdx:21]: You could drop the "Debugging" H2 here, and just close off the previous section with "The rest of this guide will cover the most common causes of webhook errors." or something along those lines. It feels like this section is recreating the ToC

**Comment by S3Prototype** [comment]
Thanks for working on this! I only did a high-level overview of this PR, as I believe that's what you were requesting in the PR description 👍.

Overall I'd say:
1. I like the location of this guide.
1. I think you should rename the headings to indicate what, overall, they're telling the user to do. "Test your webhook locally" is a good heading. I'd change "Configuration in the Dashboard" to "Check your dashboard configuration". I think more imperative language in the headings not only make it clear what the section is about, but lets the users decide at a high level whether or not they've already done what you're going to tell them to do there, so they can skip it.
1. Looks like you plan on putting screenshots of the webhook section of the dashboard in this guide. In the docs we're trying to limit screenshots to only the most absolutely necessary. An out of date screenshot can be more frustrating than none at all, and the dashboard UI could change without us realizing we need to update this image.
1. We can clean this up when it's time for a more thorough review and everything is done, but the current state of the guide feels a bit wordy

**Comment by royanger** [comment]
> Thanks for working on this! I only did a high-level overview of this PR, as I believe that's what you were requesting in the PR description 👍.
> 
> Overall I'd say:
> 
>     1. I like the location of this guide.
> 
>     2. I think you should rename the headings to indicate what, overall, they're telling the user to do. "Test your webhook locally" is a good heading. I'd change "Configuration in the Dashboard" to "Check your dashboard configuration". I think more imperative language in the headings not only make it clear what the section is about, but lets the users decide at a high level whether or not they've already done what you're going to tell them to do there, so they can skip it.
> 
>     3. Looks like you plan on putting screenshots of the webhook section of the dashboard in this guide. In the docs we're trying to limit screenshots to only the most absolutely necessary. An out of date screenshot can be more frustrating than none at all, and the dashboard UI could change without us realizing we need to update this image.
> 
>     4. We can clean this up when it's time for a more thorough review and everything is done, but the current state of the guide feels a bit wordy

I was looking at 3 images. One is a flow chart and now a screenshot. The other two are of the Svix embedded iframe. I get limiting screenshots, but customers often don't find where the errors, error codes, etc are located so I think there is value in highlighting that. Also I think the Svix embed is less likely to change versus most of our Dashboard. Let me know what you think regarding that.

File: docs/integrations/webhooks/sync-data.mdx (hunk 1 above)
Comment (alexisintech) [reply docs/integrations/webhooks/sync-data.mdx:92]:
```suggestion
Clerk's `authMiddleware()` protects all routes by default. In order for the webhook to reach your endpoint, you must make the Route Handler public. To do so, you can pass the route to the [`publicRoutes` property](/docs/references/nextjs/auth-middleware#make-routes-public-by-using-public-routes) in the `authMiddleware()` options.

The following example makes the `/api/webhooks` route and all its subroutes public:
```

File: docs/integrations/webhooks/debugging-webhooks.mdx (hunk 2 above)
Comment (alexisintech) [reply docs/integrations/webhooks/debugging-webhooks.mdx:5]: Should it be "How to debug webhooks" or "Debug webhooks" ? (For consistency with the styleguide, we don't use gerunds (-ing verb endings) in headings ([rule here](https://github.com/clerk/clerk-docs/blob/main/STYLEGUIDE.md#24-avoid-gerunds--ing-words)))

File: docs/integrations/webhooks/sync-data.mdx (hunk 1 above)
Comment (royanger) [reply docs/integrations/webhooks/sync-data.mdx:92]: @S3Prototype -- I've changed this to be inclusive, by mentioning both. They are API Routes in Pages Router and Route Handlers in App Router. There isn't a single official term that applies to both.

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 5]
```diff
@@ -0,0 +1,93 @@
+---
+title: Debugging webhooks
+description: Understand how to debug your webhooks while developing your application
+---
+
+Developing with webhooks can be a new experience for developers. It can be hard to debug when something goes awry. This guide will cover the basics for debugging and help you to direct your attention to the correct spot.
+
+## Understanding webhooks and local development
```
[the remark this answers is not in the document]
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:8]: Changed to `## Webhooks and local development`

File: docs/integrations/webhooks/debugging-webhooks.mdx (hunk 3 above)
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:65]: Added an H3, but used `### HTTP Response Codes` since these are not specific to webhooks.

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+1. The webhook route is not correctly configured
+2. Configuration in your dashboard
+3. The tunnel being used
+4. Verification of the webhook
+5. The logic you created to process the webhook
+
+
+### Testing your Webhook route locally
+
+If you are having trouble with your webhook, you can create a basic response and then test it locally. First, add the following file and code to your app. Also make sure that you add `ignoredRoutes: ['/api/webhooks(.*)']` to your `authMiddleware()`.
```
[the remark this answers is not in the document]
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:32]: I've updated this section to make it more clear it just a sample route handler not related to a webhook.

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+```bash
+curl -H 'Content-Type: application/json' \
+     -X POST http://localhost:3000/api/webhooks/test
+```
+
+If you see the `{"message":"The webhook route is working"}` then the basic webhook route is working, and ready to build on.
+
+
+### Configuration in your dashboard
```
[the remark this answers is not in the document]
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:50]: Used suggestion from Shaquill of 'Check your dashboard configuration'

**Comment by royanger** [comment]
>     3. Looks like you plan on putting screenshots of the webhook section of the dashboard in this guide. In the docs we're trying to limit screenshots to only the most absolutely necessary. 

I removed the markers for the screenshots and left only the only for the flow chart. However I would consider the screenshot showing how and where to find the responses absolutely necessary. It is common when supporting webhooks that customers have trouble finding this section.

File: docs/integrations/webhooks/debugging-webhooks.mdx (hunk 2 above)
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:5]: I changed this to `Debug your webhooks`

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+Whether you are developing locally or deploying to production, the webhook URL provided must be exact and correct. The URL breaks down into three parts -- the protocol (`http` vs `https`), the domain (`domain.com`) and the path (`/api/webhooks/user`).
+- Whether in development using a tunnel or in production, the URL will almost always use `https`.
+- The domain needs to be exact. A common error in production is not including the `www.`. Unlike entering a domain in your browser, a webhook will not be redirect from `domain.com` to `www.domain.com`. If your application lives on `www.domain.com` then the webhook URL must use that.
+- The path must match the path in your application.
```
[the remark this answers is not in the document]
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:55]: I think at this point (debugging) a customer is familiar with finding this section. I'm hesitant to go into this level of detail. I added the last-active link.

File: docs/integrations/webhooks/debugging-webhooks.mdx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+### Debugging your tunnel
+
+You've setup your route and tested it, configured your tunnel, updated your dashboard and now are trying to use the webhook but are getting errors. You will want to visit the [Webhook section](https://dashboard.clerk.com/last-active?path=webhooks) of your dashboard and the select the webhook you created.
```
[the remark this answers is not in the document]
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:59]: I would much prefer an image here. I think, as above, customer can find the general section. They struggle to find the responses when working with them in support, even when told where to find.

[the remark this answers is not in the document]
Comment (royanger) [reply docs/integrations/webhooks/debugging-webhooks.mdx:59]: Also the table was also present, but made around adjustments around it.

File: docs/integrations/webhooks/debug-your-webhooks.mdx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+1. Middleware is not configured correctly for webhooks
+2. The webhook route is not correctly configured
+3. Configuration in your dashboard
+4. The tunnel configuration
+5. Verification of the webhook
+6. The logic you created to process the webhook
+
+
+##  Check your Middleware configuration
```
Comment (alexisintech) [inline docs/integrations/webhooks/debug-your-webhooks.mdx:30]: Update for clerkMiddleware

File: docs/integrations/webhooks/debug-your-webhooks.mdx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+When you are developing on your localhost, your application is not internet facing and can't receive the webhook request. You will need to use a tool that creates a tunnel from the internet to your localhost. These tools will provide temporary or permanant URLs depending on the tool and the plan you subscribe to. Popular tools include `ngrok`, `localtunnel` and `Cloudflare Tunnel`.
+
+Below is an image demonstrating the flow from when you perform an action that triggers a webhook through until the webhook is received by your application. This flow chart will show the role of the tunnel.
+
+FLOW CHART IMAGE HERE
+
+With the Webhook flow in mind, the following areas could cause problems when trying to develop with webhooks:
+
+1. Middleware is not configured correctly for webhooks
```
Comment (alexisintech) [inline docs/integrations/webhooks/debug-your-webhooks.mdx:22]: Link these options to their corresponding sections, so `[Middleware is not configured correctly for webhooks](#check-your-middleware-configuration)`

File: docs/integrations/webhooks/debug-your-webhooks.mdx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+## Webhooks and local development
+
+When you or a user of your application performs certain actions, a webhook can be triggered. You can see the full list of [webhook events](/docs/integrations/webhooks/overview#supported-webhook-events) for a list of the actions that could result in a Webhook. Depending on the events subscribed to in the [Webhook section](https://dashboard.clerk.com/last-active?path=webhooks) of your dashboard, a webhook event will be triggered and sent to the specified endpoint in your application.
+
+When you are developing on your localhost, your application is not internet facing and can't receive the webhook request. You will need to use a tool that creates a tunnel from the internet to your localhost. These tools will provide temporary or permanant URLs depending on the tool and the plan you subscribe to. Popular tools include `ngrok`, `localtunnel` and `Cloudflare Tunnel`.
+
+Below is an image demonstrating the flow from when you perform an action that triggers a webhook through until the webhook is received by your application. This flow chart will show the role of the tunnel.
+
+FLOW CHART IMAGE HERE
```
Comment (alexisintech) [inline docs/integrations/webhooks/debug-your-webhooks.mdx:18]: Missing image

File: docs/integrations/webhooks/debug-your-webhooks.mdx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+```bash
+curl -H 'Content-Type: application/json' \
+     -X POST http://localhost:3000/api/webhooks/test
+```
+
+If you see the `{"message":"The route is working"}` then the basic Route Handler is working, and ready to build on.
+
+
+## Check your dashboard configuration
```
Comment (alexisintech) [inline docs/integrations/webhooks/debug-your-webhooks.mdx:66]: If we aren't including mentioning the "Clerk Dashboard" explicitly anywhere in this section, then we should at least put it in the title so users know we are talking about the Clerk Dashboard

```suggestion
## Check the webhook configuration in the Clerk Dashboard
```

File: docs/integrations/webhooks/debug-your-webhooks.mdx [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+| 405  | Your route is not accepting a POST request. All webhooks are POST requests, to the route must accept them. Unless you are using the route for somethinh else, you can restrict the route to POST requests only. |
+| 500  | A 500 error indicates that something failed in the webhook code. The request made it to your application, but there is a code related issue. This is likely a webhook verification issue or something in the webhook logic. Please see the following sections.
+
+You may run into other errors. See [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) from MDN for reference.
+
+
+## Debug webhook verfication
+
+To verify the webhook, your webhook route will need the Signing Secret from the Webhook section of the dashboard.  It will also need the Svix headers which are included in the request. Please see [Create the endpoint in your application](/docs/integrations/webhooks/sync-data#create-the-endpoint-in-your-application) for a detailed code example. You can also visit the [guide](https://docs.svix.com/receiving/verifying-payloads/how) on verifying webhooks with the Svix library.
```
Comment (alexisintech) [inline docs/integrations/webhooks/debug-your-webhooks.mdx:96]:
```suggestion
To verify the webhook, your webhook route will need the **Signing Secret** from the Clerk Dashboard and the Svix headers from the payload. See the [Create a webhook](/docs/integrations/webhooks/create-a-webhook#create-a-webhook-handler) guide for a detailed code example. You can also visit the [Svix guide](https://docs.svix.com/receiving/verifying-payloads/how) on verifying webhooks.
```

File: docs/integrations/webhooks/sync-data.mdx
Comment (alexisintech) [inline docs/integrations/webhooks/sync-data.mdx:1]: so this guide and the "create a webhook" guide are almost identical - do we want to merge the two? or do we want to keep this one as a standalone guide, and add a working example where we sync the user table to a third-party's user table (because this is the question that gets asked a lot in regards to webhooks)

File: docs/integrations/webhooks/debug-your-webhooks.mdx (hunk 12 above)
Comment (royanger) [reply docs/integrations/webhooks/debug-your-webhooks.mdx:18]: I sent you this image before via DMs for feedback/etc.

File: docs/integrations/webhooks/sync-data.mdx
Comment (royanger) [reply docs/integrations/webhooks/sync-data.mdx:1]: I think there is some overlap, but not a ton. That is partly why I tried to avoid any sort of 'this is how you set it up' in this guide, to keep it lean and focused on troubleshooting. I also suspect this will get expanded. There is also the search/name value of something focused on troubleshooting versus creating.

Comment (alexisintech) [reply docs/integrations/webhooks/sync-data.mdx:1]: Sorry, I'm talking about the "sync-data" guide!!
I think the sync data guide should sit on top of the "Create a webhook" guide - like in the intro, it could say "You should have followed this guide to create a webhook" and then it gets into the details of what specific endpoints they need to set up to sync our userbase with their third-party userbase.

Comment (royanger) [reply docs/integrations/webhooks/sync-data.mdx:1]: Honestly, I would just expanded the Sync guide with data from your new Create guide. The sync guide already covered a bunch of that material. I figured there was a reason for creating the new guide so I didn't ask about it.

File: docs/integrations/webhooks/debug-your-webhooks.mdx [hunk 15]
```diff
@@ -13,23 +13,31 @@ When you or a user of your application performs certain actions, a webhook can b
...
+3. [Configuration in your dashboard](#check-your-configuration-in-the-clerk-dashboard)
+4. [The tunnel configuration](#debug-your-tunnel-and-webhook-delivery)
+5. [Verification of the webhook](#debug-webhook-verification)
+6. [The logic you created to process the webhook](#check-your-logic)
 
 
 ##  Check your Middleware configuration
 
-Incoming webhook events will never be signed in -- they are coming from a source outside of your application. Since they will be in a signed out state, the simplest configuration is to use `ignoredRoutes` for your webhook routes. Below is an example of making all webhook routes ignored.
+Incoming webhook events will never be signed in -- they are coming from a source outside of your application. Since they will be in a signed out state, the route should be public. The simplest configuration is to use `ignoredRoutes` for your webhook routes when using `authMiddleware` and applying no protection to the route with `clerkMiddleware`. Below is an example of making all webhook routes ignored.
```
Comment (alexisintech) [inline docs/integrations/webhooks/debug-your-webhooks.mdx:28]:
```suggestion
Incoming webhook events will never be signed in -- they are coming from a source outside of your application. Since they will be in a signed out state, the route should be public. 

The following example shows the recommended Middleware configuration for your webhook routes.
```

**Review by alexisintech — APPROVED** [review]
